// @cfcs/core/src/reactive/Observer.ts
import Component from "@egjs/component";
import { getCurrentDetected } from "./detectDependencies";

interface EmitterEvents<Value> {
  update: (value: Value, prevValue: Value) => void;
}

/**
 * @memberof Reactive
 */
export class Observer<Value = any> {
  protected _current: Value;
  protected _emitter = new Component<EmitterEvents<Value>>();

  /**
   *
   */
  constructor(value?: Value) {
    this._current = value as any;
  }
  public get current() {
    const currentDetected = getCurrentDetected();

    currentDetected?.push(this);
    return this._current as Value;
  }
  public set current(value: Value) {
    this._setCurrent(value);
  }
  public subscribe(callback: (value: Value, prevValue: Value) => void) {
    this.current;
    this._emitter.on("update", callback);
    return this;
  }
  public unsubscribe(callback?: (value: Value, prevValue: Value) => void) {
    this._emitter.off("update", callback);
    return this;
  }
  protected _setCurrent(value: Value) {
    const prevValue = this._current;
    const isUpdate = value !== prevValue;

    this._current = value;

    if (isUpdate) {
      this._emitter.trigger("update", value, prevValue);
    }
  }
  public toString() {
    return `${this.current}`;
  }
  public valueOf() {
    return this.current;
  }
}
